File: Sources/ExcalidrawGeometry/HitTest.swift
import ExcalidrawMath
import ExcalidrawModel
import Foundation

/// Pointer hit-testing for elements, porting the entry logic from
/// `packages/element/src/collision.ts` (`hitElementItself`, `shouldTestInside`,
/// `isPointInElement`) and `distance.ts` (`distanceToElement`).
///
/// Scope note: shapes are tested against straight-edged outlines (rounded
/// corners are treated as sharp) and curved arrows against their point polyline.
/// This is faithful for the common cases; exact rounded-corner / cubic-Bézier
/// hit-testing arrives with the rough.js shape port.
public enum HitTest {
    /// Whether the element should be hit "from the inside" (filled / textful /
    /// closed) rather than only on its outline (`shouldTestInside`).
    public static func shouldTestInside(_ element: ExcalidrawElement) -> Bool {
        if case .arrow = element.kind { return false }

        let draggableFromInside =
            (hasBackground(element.kind) && !isTransparent(element.base.backgroundColor))
            || hasBoundText(element)
            || isIframeLike(element.kind)
            || isText(element.kind)

        switch element.kind {
        case let .line(props):
            return draggableFromInside && ElementGeometry.isPathALoop(props.points)
        case let .freedraw(props):
            return draggableFromInside && ElementGeometry.isPathALoop(props.points)
        case .image:
            return true
        default:
            return draggableFromInside
        }
    }

    /// Whether `point` (scene coords) hits the element within `threshold`
    /// (`hitElementItself`).
    public static func hit(_ element: ExcalidrawElement, at point: Point, threshold: Double) -> Bool {
        // Cheap early-out against the rotated bounding box.
        let nonRotated = ElementGeometry.bounds(element, nonRotated: true)
        guard isPointInRotatedBounds(point, nonRotated, angle: element.base.angle, tolerance: threshold) else {
            return false
        }
        if shouldTestInside(element) {
            return isPointInside(element, point: point) || isPointOnOutline(element, point: point, threshold: threshold)
        }
        return isPointOnOutline(element, point: point, threshold: threshold)
    }

    /// Strict interior test (`isPointInElement`). Open paths have no interior.
    public static func isPointInside(_ element: ExcalidrawElement, point: Point) -> Bool {
        let (x1, y1, x2, y2, cx, cy) = ElementGeometry.absoluteCoords(element)
        if case .ellipse = element.kind {
            let local = point.rotated(around: Point(cx, cy), by: -element.base.angle)
            return Ellipse(center: Point(cx, cy), halfWidth: (x2 - x1) / 2, halfHeight: (y2 - y1) / 2).includes(local)
        }
        let outline = ElementGeometry.unrotatedOutline(element)
        guard outline.closed, outline.points.count >= 3 else { return false }
        let rotated = outline.points.map { $0.rotated(around: Point(cx, cy), by: element.base.angle) }
        return Polygon(rotated).includes(point)
    }

    /// Whether `point` lies on the element outline within `threshold`
    /// (`isPointOnElementOutline`).
    public static func isPointOnOutline(_ element: ExcalidrawElement, point: Point, threshold: Double) -> Bool {
        distance(element, to: point) <= Swift.max(threshold, ExcalidrawMath.precision)
    }

    /// Shortest distance from `point` to the element outline (`distanceToElement`).
    public static func distance(_ element: ExcalidrawElement, to point: Point) -> Double {
        let (x1, y1, x2, y2, cx, cy) = ElementGeometry.absoluteCoords(element)
        let center = Point(cx, cy)
        if case .ellipse = element.kind {
            let local = point.rotated(around: center, by: -element.base.angle)
            return Ellipse(center: center, halfWidth: (x2 - x1) / 2, halfHeight: (y2 - y1) / 2).distance(from: local)
        }
        let segments = rotatedSegments(element, center: center)
        guard let first = segments.first else {
            // Degenerate (single point) element.
            let outline = ElementGeometry.unrotatedOutline(element)
            return outline.points.first.map {
                $0.rotated(around: center, by: element.base.angle).distance(to: point)
            } ?? Swift.min(abs(point.x - x1), abs(point.y - y1))
        }
        return segments.dropFirst().reduce(first.distance(to: point)) { acc, segment in
            Swift.min(acc, segment.distance(to: point))
        }
    }

    // MARK: Internals

    static func rotatedSegments(_ element: ExcalidrawElement, center: Point) -> [LineSegment] {
        let outline = ElementGeometry.unrotatedOutline(element)
        let pts = outline.points.map { $0.rotated(around: center, by: element.base.angle) }
        guard pts.count >= 2 else { return [] }
        var segments: [LineSegment] = []
        for i in 0..<(pts.count - 1) {
            segments.append(LineSegment(pts[i], pts[i + 1]))
        }
        if outline.closed, let first = pts.first, let last = pts.last, first != last {
            segments.append(LineSegment(last, first))
        }
        return segments
    }

    static func isPointInRotatedBounds(
        _ point: Point, _ bounds: BoundingBox, angle: Double, tolerance: Double
    ) -> Bool {
        let center = Point((bounds.minX + bounds.maxX) / 2, (bounds.minY + bounds.maxY) / 2)
        let adjusted = angle == 0 ? point : point.rotated(around: center, by: -angle)
        return adjusted.isWithin(
            Point(bounds.minX - tolerance, bounds.minY - tolerance),
            Point(bounds.maxX + tolerance, bounds.maxY + tolerance)
        )
    }

    private static func hasBackground(_ kind: ElementKind) -> Bool {
        switch kind {
        case .rectangle, .iframe, .embeddable, .ellipse, .diamond, .line, .freedraw: return true
        default: return false
        }
    }

    private static func isTransparent(_ color: String) -> Bool {
        if color == "transparent" || color.isEmpty { return true }
        // 8-digit hex with a fully transparent alpha channel.
        return color.count == 9 && color.hasPrefix("#") && color.hasSuffix("00")
    }

    private static func hasBoundText(_ element: ExcalidrawElement) -> Bool {
        element.base.boundElements?.contains { $0.type == .text } ?? false
    }

    private static func isIframeLike(_ kind: ElementKind) -> Bool {
        switch kind {
        case .iframe, .embeddable: return true
        default: return false
        }
    }

    private static func isText(_ kind: ElementKind) -> Bool {
        if case .text = kind { return true }
        return false
    }
}

File: web/packages/excalidraw-svelte/src/geometry/hit-test.ts
import { Ellipse, LineSegment, PRECISION, Point, Polygon } from "../math/index.js";
import type { ExcalidrawElement } from "../model/index.js";
import type { BoundingBox } from "./bounding-box.js";
import { absoluteCoords, bounds, isPathALoop, unrotatedOutline } from "./element-geometry.js";

function isTransparent(color: string): boolean {
  if (color === "transparent" || color.length === 0) return true;
  return color.length === 9 && color.startsWith("#") && color.endsWith("00");
}

function hasBackground(type: ExcalidrawElement["type"]): boolean {
  switch (type) {
    case "rectangle":
    case "iframe":
    case "embeddable":
    case "ellipse":
    case "diamond":
    case "line":
    case "freedraw":
      return true;
    default:
      return false;
  }
}

function hasBoundText(el: ExcalidrawElement): boolean {
  return el.boundElements?.some((b) => b.type === "text") ?? false;
}

/** Whether the element should be hit "from the inside" (`shouldTestInside`). */
export function shouldTestInside(el: ExcalidrawElement): boolean {
  if (el.type === "arrow") return false;
  const draggableFromInside =
    (hasBackground(el.type) && !isTransparent(el.backgroundColor)) ||
    hasBoundText(el) ||
    el.type === "iframe" ||
    el.type === "embeddable" ||
    el.type === "text";

  if (el.type === "line" || el.type === "freedraw") {
    return draggableFromInside && isPathALoop(el.points);
  }
  if (el.type === "image") return true;
  return draggableFromInside;
}

function rotatedSegments(el: ExcalidrawElement, center: Point): LineSegment[] {
  const outline = unrotatedOutline(el);
  const pts = outline.points.map((p) => p.rotated(center, el.angle));
  if (pts.length < 2) return [];
  const segments: LineSegment[] = [];
  for (let i = 0; i < pts.length - 1; i++) segments.push(new LineSegment(pts[i]!, pts[i + 1]!));
  const first = pts[0]!;
  const last = pts[pts.length - 1]!;
  if (outline.closed && !first.equals(last)) segments.push(new LineSegment(last, first));
  return segments;
}

function isPointInRotatedBounds(
  point: Point,
  box: BoundingBox,
  angle: number,
  tolerance: number,
): boolean {
  const center = new Point((box.minX + box.maxX) / 2, (box.minY + box.maxY) / 2);
  const adjusted = angle === 0 ? point : point.rotated(center, -angle);
  return adjusted.isWithin(
    new Point(box.minX - tolerance, box.minY - tolerance),
    new Point(box.maxX + tolerance, box.maxY + tolerance),
  );
}

/** Strict interior test (`isPointInElement`). Open paths have no interior. */
export function isPointInside(el: ExcalidrawElement, point: Point): boolean {
  const { x1, y1, x2, y2, cx, cy } = absoluteCoords(el);
  const center = new Point(cx, cy);
  if (el.type === "ellipse") {
    const local = point.rotated(center, -el.angle);
    return new Ellipse(center, (x2 - x1) / 2, (y2 - y1) / 2).includes(local);
  }
  const outline = unrotatedOutline(el);
  if (!outline.closed || outline.points.length < 3) return false;
  const rotated = outline.points.map((p) => p.rotated(center, el.angle));
  return new Polygon(rotated).includes(point);
}

/** Shortest distance from `point` to the element outline (`distanceToElement`). */
export function distanceToElement(el: ExcalidrawElement, point: Point): number {
  const { x1, y1, x2, y2, cx, cy } = absoluteCoords(el);
  const center = new Point(cx, cy);
  if (el.type === "ellipse") {
    const local = point.rotated(center, -el.angle);
    return new Ellipse(center, (x2 - x1) / 2, (y2 - y1) / 2).distance(local);
  }
  const segments = rotatedSegments(el, center);
  const first = segments[0];
  if (first === undefined) {
    const outline = unrotatedOutline(el);
    const p0 = outline.points[0];
    return p0 !== undefined
      ? p0.rotated(center, el.angle).distance(point)
      : Math.min(Math.abs(point.x - x1), Math.abs(point.y - y1));
  }
  let min = first.distanceToPoint(point);
  for (let i = 1; i < segments.length; i++) {
    min = Math.min(min, segments[i]!.distanceToPoint(point));
  }
  return min;
}

/** Whether `point` lies on the element outline within `threshold`. */
export function isPointOnOutline(el: ExcalidrawElement, point: Point, threshold: number): boolean {
  return distanceToElement(el, point) <= Math.max(threshold, PRECISION);
}

/** Whether `point` lies within the element's rotated bounding box (`hitElementBoundingBox`). */
export function isPointInElementBounds(
  el: ExcalidrawElement,
  point: Point,
  tolerance = 0,
): boolean {
  return isPointInRotatedBounds(point, bounds(el, true), el.angle, tolerance);
}

/** Whether `point` (scene coords) hits the element within `threshold` (`hitElementItself`). */
export function hit(el: ExcalidrawElement, point: Point, threshold: number): boolean {
  const nonRotated = bounds(el, true);
  if (!isPointInRotatedBounds(point, nonRotated, el.angle, threshold)) return false;
  if (shouldTestInside(el)) {
    return isPointInside(el, point) || isPointOnOutline(el, point, threshold);
  }
  return isPointOnOutline(el, point, threshold);
}
